### PR TITLE
Error handling

### PR DESCRIPTION
**Pull request by barnettwilliam**
Common handling of errors and exceptions via ErrorHandler.

Unexpected error notifcation are displayed for any unhandled exceptions via `window.onerror()`  that ErrorHandler initialises to call `notify()`.

### CODE REVIEW

File: platform/src/EducationPlatformApp.js [hunk 1]
```diff
@@ -152,7 +155,7 @@ class EducationPlatformApp {
 
         if (errors.length==0){
             // An activity configuration has been provided
-            this.toolsManager = new ToolsManager(this.errorNotification);
+            this.toolsManager = new ToolsManager(this.errorHandler.notify.bind(this.errorHandler));
```
Comment (szschaler) [inline platform/src/EducationPlatformApp.js:158]: This is quite complicated "magic". Is there a way of simplifying this?

File: platform/src/EducationPlatformApp.js [hunk 2]
```diff
@@ -576,59 +546,63 @@ class EducationPlatformApp {
...
         } else {
-            //Buttons defined by tool
-            buttonConfig= action.source.ref.buttons.find( btn => btn.id == sourceButton );
-        }  
+            // Action found so try and invoke
+            let buttonConfig;
+            
+            if(action.source.buttons){
+                //Buttons defined by activity
+                buttonConfig=  action.source.buttons.find( btn => btn.id == sourceButton );
```
Comment (szschaler) [inline platform/src/EducationPlatformApp.js:560]:
```suggestion
                buttonConfig = action.source.buttons.find (btn => btn.id == sourceButton);
```

(I know this isn't part of this PR, but might as well clean it up while we're here.)

File: platform/src/EducationPlatformApp.js [hunk 3]
```diff
@@ -576,59 +546,63 @@ class EducationPlatformApp {
...
+            } else {
+                //Buttons defined by tool
+                buttonConfig= action.source.ref.buttons.find( btn => btn.id == sourceButton );
```
Comment (szschaler) [inline platform/src/EducationPlatformApp.js:563]:
```suggestion
                buttonConfig = action.source.ref.buttons.find (btn => btn.id == sourceButton);
```
ditto

**Review by szschaler — APPROVED** [review]
Generally looks good, just some small suggestions.

File: platform/src/EducationPlatformApp.js (hunk 1 above)
Comment (barnettwilliam) [reply platform/src/EducationPlatformApp.js:158]: @szschaler the `errorNotification()` etc could be moved to `PlaygroundUtility.js` and imported as required from there and means that it doesn't have to be passed into `ErrorHandler`

Comment (barnettwilliam) [reply platform/src/EducationPlatformApp.js:158]: Simplified by moving notification functions to `PlaygroundUtility` in bbabad1ec6f1ea7a340b766d1c5ce449b0c08ed7. 

Rename `PlaygroundUtility` as a separate task

File: platform/test/spec/testEducationPlatformAppSpec.js [hunk 4]
```diff
@@ -169,91 +169,4 @@ describe("EducationPlatformApp", () => {
             })
         })
     })
-
-
-    describe("notification()", () => {
```
Comment (szschaler) [inline platform/test/spec/testEducationPlatformAppSpec.js:140]: What's happened to all of these tests? Wouldn't it make sense to move them to the PlatformUtil tests spec?

File: platform/test/spec/testErrorHandlerSpec.js [hunk 5]
```diff
@@ -1,27 +1,20 @@
...
+import { PlaygroundUtility } from "../../src/PlaygroundUtility.js";
 
 describe("ErrorHandler", () => {
 
     // Setup
     let notifierSpy;
 
     beforeEach(()=>{
-        notifierSpy = jasmine.createSpy("notifier");
+        notifierSpy = spyOn(PlaygroundUtility,"errorNotification");
```
Comment (szschaler) [inline platform/test/spec/testErrorHandlerSpec.js:14]: This is still being passed in as a parameter to the `ErrorHandler` constructor calls below. Doesn't that need fixing, too?

**Review by szschaler — CHANGES_REQUESTED** [review]

File: platform/test/spec/testEducationPlatformAppSpec.js (hunk 4 above)
Comment (barnettwilliam) [reply platform/test/spec/testEducationPlatformAppSpec.js:140]: Yes, these were moved but the file was missing from the commit.  Fixed in b4d4f4a293d734496827ac955323ea46d437b208

File: platform/test/spec/testErrorHandlerSpec.js (hunk 5 above)
Comment (barnettwilliam) [reply platform/test/spec/testErrorHandlerSpec.js:14]: Thanks yes, fixed in 27027ebcef16eba9977cddc6e63726d01cdae3a0

**Review by szschaler — APPROVED** [review]

**Pull request merged**